qutip/qobjevo.py

```diff
@@ -646,12 +623,14 @@ def _check_old_with_state(self):
 
     def __del__(self):
         for file_ in self.coeff_files:
-            try:
-                os.remove(file_)
-            except:
-                pass
-
-    def __call__(self, t, data=False, state=None, args={}):
+            # Using os module to manipulate filesystem is not thread-safe,
+            # seems fine at the moment though. For the same reason,
+            # do not ignore exceptions here.
+            os.remove(file_)
```

Comment:
File-system manipulation is inherently thread-unsafe since you're mutating global state accessible from many locations, so I'm not sure what you mean by this.  The `__del__` destructor of a Python object will only be called by one thread, with safety maintained by the GIL, so provided we enforce that these files are internal to the `QobjEvo` instance (which we do), I don't see a threading issue with this.

Also, I don't understand what you mean by "for the same reason, do not ignore exceptions here".  The major possible exception is that the file does not exist already (perhaps the user cleaned it up manually?).  In this case, ignoring the exception will guarantee that no later file will be deleted.  It should not be an unhandled exception for a file to fail to be deleted, even if it's a permission or other access issue; at worst we've just littered the user's directory with a little bit of detritus, whereas an unhandled exception crashes their script.

Reply:
First of all, ``os.remove()`` uses relative path at the moment, and ``working path`` is a global state affecting all threads, if any thread modifies current working path, all ``os.remove()`` related operations will be sabotaged. I think it is best to use absolute path with ``pathlib``. 
Since we do enforce that those files are internal, it shouldn't be the case where those files were deleted before we delete a ``QobjEvo`` (it is very unlikely that users can remove the file before our code does and not causing a problem in importing corresponding module i.e. ``ImportError``). If files were deleted before destructor of ``QobjEvo`` kicks in, either there are some bugs in our code or users deleting those files will get into trouble at some point. To sum up, I don't think covering up potential bugs in our code by ignoring impossible but 'harmless' edge cases.

Reply:
The point about absolute paths is good - I assumed that the files were all absolute already.  This change here does not change any behaviour to do with that, though.

If you are most concerned about ensuring you clean things up, you should not be throwing unhandled exceptions until all your cleanup code has had a chance to run.  If you are particularly concerned about this, then you could run the entire loop with a `try: pass; except OSError` on each step and re-raise afterwards with a merged error.

There _are_ several possibilities for an `OSError` to be raised when trying to delete a file which do show that something is a bit wrong, but are not errors in the rest of the programme.  In a long-running script, the user may already have tidied up the files.  This is not impossible and will not generally cause an error at runtime; `importlib` caches imported modules, so will not search for the file, and the file will certainly have been imported before the `__del__` method is called.  If the user tries to `importlib.reload` (on a completely private module they "officially" have no way of knowing about), _then_ they will get the `ModuleNotFoundError` which is a fine way to handle it.  If the permissions on the directory are changed while the script is running, the file may also fail to be deleted, which is again weird but not an error and should not raise - this can happen when running on multi-process clusters which may dispatch to a different node.  These are unlikely edge cases, but they are _not_ bugs.  Yes, somewhere in QuTiP deleting the file before the `__del__` method is likely a mistake, but in this case there is no harm done by the file failing to be deleted (and if it causes issues, we will get an `ImportError` elsewhere), whereas there is harm done by raising too early when cleaning a list of files where only one failed because the user had opened it in a text editor (thanks Windows) or raising because you're running code on a different cluster node which doesn't have write permissions to the work directory.

We're not looking to punish users.  Methods should not raise unless we cannot continue safely and unambiguously.

The file handling in `QobjEvo` is not perfect, you're right.  Right now, it's essentially temporary code; the whole time-dependent coefficient handling is changed entirely in the next major branch, and none of this code exists any more.  We'd be happy to merge bug-fixes for an identifiable problem in the short-term, but there's little point in changing working code that is scheduled for deletion.